xgi/algorithms/assortativity.py

```diff
@@ -0,0 +1,156 @@
+import random
+from itertools import combinations
+
+import numpy as np
+
+import xgi
+from xgi.exception import XGIError
+
+__all__ = ["dynamical_assortativity", "degree_assortativity"]
+
+
+def dynamical_assortativity(H):
+    """Computes the dynamical assortativity of a uniform hypergraph.
+
+    Parameters
+    ----------
+    H : xgi.Hypergraph
+        Hypergraph of interest
+
+    Returns
+    -------
+    float
+        The dynamical assortativity
+
+    Raises
+    ------
+    XGIError
+        If the hypergraph is not uniform, or if there are no nodes
+        or no edges
+
+    References
+    ----------
+    Nicholas Landry and Juan G. Restrepo,
+    Hypergraph assortativity: A dynamical systems perspective,
+    Chaos 2022.
+    DOI: 10.1063/5.0086905
+
+    """
+    if not xgi.is_uniform(H):
+        raise XGIError("Hypergraph must be uniform!")
+
+    if H.num_nodes == 0 or H.num_edges == 0:
+        raise XGIError("Hypergraph must contain nodes and edges!")
+
+    degs = H.degree()
+    k1 = np.mean([degs[n] for n in H.nodes])
+    k2 = np.mean([degs[n] ** 2 for n in H.nodes])
+    kk1 = np.mean(
+        [
+            degs[n1] * degs[n2]
+            for e in H.edges
+            for n1, n2 in combinations(H.edges.members(e), 2)
+        ]
+    )
+
+    return kk1 * k1**2 / k2**2 - 1
+
+
+def degree_assortativity(H, type="uniform", exact=False, num_samples=1000):
+    """Computes the degree assortativity of a hypergraph
+
+    Parameters
+    ----------
+    H : Hypergraph
+        The hypergraph of interest
+    type : str, default: "uniform"
+        the type of degree assortativity. valid choices are
+        "uniform", "top-2", and "top-bottom".
+    exact : bool, default: False
+        whether to compute over all edges or
+        sample randomly from the set of edges
+    num_samples : int, default: 1000
+        if not exact, specify the number of samples for the computation.
+
+    Returns
+    -------
+    float
+        the degree assortativity
+
+    References
+    ----------
+    Phil Chodrow,
+    Configuration models of random hypergraphs,
+    Journal of Complex Networks 2020.
+    DOI: 10.1093/comnet/cnaa018
+    """
+    degs = H.degree()
+    if exact:
+        k1k2 = [
+            choose_degrees(e, degs, type)
+            for e in H.edges
+            if len(H.edges.members(e)) > 1
+        ]
```

Comment:
Just out of curiosity (and if you have time), could you please check whether the following is faster?
```python
[choose_degrees(e, degs, type) for e in H.edges.filterby("order", 1, "geq")]
```

Reply:
The method that is currently in place is ~25% faster than that.

Reply:
I could replace it for readability though

Reply:
:sob: :sob: :sob: no, leave it, I'll have to profile the stats package..